cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+            {
+                Log(LOG_LEVEL_VERBOSE, "have error: %s",
+                    line + strlen("Error="));
+                have_error = true;
+            }
+            else if (StringStartsWith(line, "ErrorMessage="))
+            {
+                Log(LOG_LEVEL_VERBOSE, "have error message: %s",
+                    line + strlen("ErrorMessage="));
+                have_error = true;
+            }
+            else
+            {
+                 Log(LOG_LEVEL_INFO,
+                     "Unsupported response received form package module: %s",
+                     line);
+                 have_error = true;
+            }
+        }
+        /* We have one more entry left. */
+        if (package_data[0] && package_data[1] && package_data[2])
+        {
+            WritePackageDataToDB(db_cached, package_data[0],
+                             package_data[1], package_data[2], type);
+            
+            char inventory_line[strlen(package_data[0]) +
+                                strlen(package_data[1]) +
+                                strlen(package_data[2]) + 4];
+
+            xsnprintf(inventory_line, sizeof(inventory_line),
+                      "%s,%s,%s\n", package_data[0], package_data[1],
+                      package_data[2]);
+            BufferAppendString(inventory_data, inventory_line);
+        }
+        
+        char *inventory_key = "<inventory>";
+        char *inventory_list = BufferClose(inventory_data);
+        
+        /* We can have empty list of installed software or available updates. */
+        if (!inventory_list)
+        {
+            WriteDB(db_cached, inventory_key, "\n", 1);
+        }
+        else
+        {
+            WriteDB(db_cached, inventory_key, inventory_list,
+                    strlen(inventory_list));
+            free(inventory_list);
+        }
+        
+        CloseDB(db_cached);
+        return have_error ? -1 : 0;
+    }
+    /* Unable to open database. */
+    return -1;
+}
+
+
+bool UpdateCache(Rlist* options, const PackageModuleWrapper *wrapper,
+                 UpdateType type)
+{
+    char *options_str = ParseOptions(options);
+    Rlist *response = NULL;
+    
+    const char *req_type = NULL;
+    if (type == UPDATE_TYPE_INSTALLED)
+    {
+        req_type = "list-installed";
+    }
+    else if (type == UPDATE_TYPE_UPDATES)
+    {
+        req_type = "list-updates";
+    }
+    else if (type == UPDATE_TYPE_LOCAL_UPDATES)
+    {
+        req_type = "list-updates-local";
+    }
+    if (ReadWriteDataToPackageScript(req_type, options_str, &response,
+            wrapper) != 0)
+    {
+        Log(LOG_LEVEL_VERBOSE, "Some error occurred while communicating with "
+                "package module while updating cache.");
+        free(options_str);
+        return false;
+    }
+    
+    if (!response)
+    {
+        Log(LOG_LEVEL_DEBUG,
+            "Received empty packages list after requesting: %s", req_type);
+    }
```

Comment:
Should the `UpdatePackagesDB` call be bypassed if `!response` ?  How does it cope with being passed a NULL response ?